src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java

```diff
@@ -1046,11 +1075,22 @@ protected ThreadInvocationCounter initialValue() {
 
         this.configPrinterRegistration = bundleContext.registerService(Object.class.getName(),
                 new ModelConfigurationPrinter(this, bundleContext, adapterImplementations), printerProps);
+
+        Hashtable<Object, Object> listenerProps = new Hashtable<>();
```

Comment:
What about registering this service via DS? Seems easier and requires less code IMHO.

Reply:
The problem with DS in this case (which is already true of the `ConfigurationPrinter` right above this) is that it needs access to internal data structures of the `ModelAdapterFactory`, so I'd rather keep that all internal and manage the lifecycle manually.

Reply:
If I understand correctly all service properties are constants. If you add the ServletRequestListener to the Service annotation from above, the same component (i.e. object/instance) would be used for both the ModelAdapterFactory and the ServletRequestListener so you would not have any problems access internal data structure of ModelAdapterFactory from ServletRequestListener.onDispose. The only drawback I see is that you need to combine/expose the properties for both services (i.e. ModelAdapterFactory and ServletRequestListener) in this single component.

Reply:
Ah, I see what you mean now. Yeah, that might work.